presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/InMemoryHashAggregationBuilder.java

```diff
@@ -152,27 +233,82 @@ protected Page computeNext()
                         BlockBuilder output = pageBuilder.getBlockBuilder(types.size() + i);
                         aggregator.evaluate(groupId, output);
                     }
-
-                    groupId++;
                 }
 
                 return pageBuilder.build();
             }
         };
     }
 
+    public List<Type> buildTypes()
+    {
+        ArrayList<Type> types = new ArrayList<>(groupByHash.getTypes());
+        for (Aggregator aggregator : aggregators) {
+            types.add(aggregator.getType());
+        }
+        return types;
+    }
+
+    private Iterator<Integer> consecutiveGroupIds()
+    {
+        return IntStream.range(0, groupByHash.getGroupCount()).iterator();
+    }
+
+    private Iterator<Integer> hashSortedGroupIds()
```

Comment:
I think we should avoid all the boxing this is going to add. `fastutil` has a `IntIterator` class, as well as sorting methods that operate on primitives